src/librustc_trans/assert_module_sources.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This pass is only used for UNIT TESTS related to incremental
+//! compilation. It tests whether a particular `.o` file will be re-used
+//! from a previous compilation or whether it must be regenerated.
+//!
+//! The user adds annotations to the crate of the following form:
+//!
+//! ```
+//! #![rustc_partition_reused(module="spike", cfg="rpass2")]
+//! #![rustc_partition_translated(module="spike-x", cfg="rpass2")]
```

Comment:
> Yes that, but also logically it's another case, right? 

Hmm, well, logically yes, though (at least now) we don't create a `ModuleTranslation` for modules that used to exist in some previous compilation but no longer do, since there is no work to do in such a case -- but I guess we could just assert that (indeed) no module exists with that name.

> Isn't this covered by reconcile_work_products() anyway?

Yeah, I forgot about that. Indeed, for the most part it is, which is why I delete the other code I had that just scraped up files with "suspicious looking" names.